algorithms/sorting/strand_sort.py

```diff
@@ -17,53 +17,46 @@
     Psuedo Code: https://en.wikipedia.org/wiki/Strand_sort
 """
 
-from collections import deque
 
-
-def sort(array):
-    if len(array) < 2:
-        return array
+def sort(unsorted):
+    if len(unsorted) < 2:
+        return unsorted
     result = []
-    while array:
-        sublist = [array.pop()]
+    while unsorted:
+        sublist = [unsorted.pop(0)]
+        leftovers = []
         last = sublist[0]
-        sub_append = sublist.append
-        leftovers = deque()
-        left_append = leftovers.append
-        for item in array:
+        sublist_append = sublist.append
+        leftovers_append = leftovers.append

```

Comment:
Oh, I misunderstood what you'd asked originally. My apologies.

I assign both `append`s to variables because, with higher numbers of calls, function calls are [slower][0]... ish. They're not crippling, but they're noticeably slower when handling larger and larger sets of calls. The [reason][1] is because every time the function is called, the `LOAD_GLOBAL` opcode has to look for the function name in the object calling that function (`LOAD_GLOBAL -> `LOAD_ATTR` -> `LOAD_CONST`), which is one step more than calling it when assigned to a variable (`LOAD_GLOBAL -> `LOAD_CONST`).

    >>> import dis
    >>> a = [1,2,3]
    >>> def foo():
    ...     a.append(1)
    ...
    >>> dis.dis(foo)
      2           0 LOAD_GLOBAL              0 (a)
                  3 LOAD_ATTR                1 (append)
                  6 LOAD_CONST               1 (1)
                  9 CALL_FUNCTION            1 (1 positional, 0 keyword pair)
                 12 POP_TOP
                 13 LOAD_CONST               0 (None)
                 16 RETURN_VALUE
    >>> a_append = a.append
    >>> def foo():
    ...     a_append(1)
    ...
    >>> dis.dis(foo)
      2           0 LOAD_GLOBAL              0 (a_append)
                  3 LOAD_CONST               1 (1)
                  6 CALL_FUNCTION            1 (1 positional, 0 keyword pair)
                  9 POP_TOP
                 10 LOAD_CONST               0 (None)
                 13 RETURN_VALUE

This is not a massive slowdown, but when going for speed, it's necessary and somewhat elegant.

To show the difference:

    100000 already sorted       Mean                    Min                     Max
    original commited version:  20.337856443215422      19.674197121421525      21.00151576500932
    sublist/leftover . append:  0.03793888725704164     0.03603723264198777     0.039840541872095514
    sublist/leftover _ append:  0.03085112794449074     0.030596452507197114    0.03110580338178437

    100000 reverse sorted       Mean                    Min                     Max
    original commited version:  21.95350015622824       21.846741851263694      22.060258461192788
    sublist/leftover . append:  18.66631388535538       18.25951236894673       19.073115401764028
    sublist/leftover _ append:  15.523921059683886      15.237184018143552      15.81065810122422

    100000 random order         Mean                    Min                     Max
    original commited version:  18.43586113942463       18.334801800779132      18.53692047807013
    sublist/leftover . append:  14.09712203590557       14.05387709342915       14.14036697838199
    sublist/leftover _ append:  12.889577499154484      12.73960818893434       13.039546809374627

I realize that this is a learning library, so speed might not be a big desire or care, but I also think it's good to have a performant implementation. (I might be biased, but I'm still offended at the original version of the Python Strand Sort on Wikipedia. \*shudders*)

In the end, it's your library, so lemme know what you want! As you can see, the differences aren't massive, so changing it back is easy-peasy and loses little.

  [0]: https://wiki.python.org/moin/PythonSpeed/PerformanceTips#Avoiding_dots...
  [1]: http://stackoverflow.com/a/14648595/3023252